### ISSUE TITLE
Kafka consumer doesn't use context from ConsumerRecord Headers

### ISSUE THREAD

**Issue by gbiv** [DataDog/dd-trace-java#3397]
I'm using [Alpakka](https://doc.akka.io/docs/alpakka-kafka/current/index.html) in a scala service to subscribe to topics.
I'm seeing that the `ConsumerRecord`s that are being processed in my `Flow` have the correct header values for `x-datadog-trace-id` and `x-datadog-parent-id` however values for `mdc.dd.span_id` and `mdc.dd.trace_id` are completely different than those in the headers for the same call.
![Screen Shot 2022-02-16 at 1 09 15 PM](https://user-images.githubusercontent.com/37077586/154343916-32ea12b8-4f1b-4bda-97c2-47959406aa57.png)

This causes any calls to other services, via grpc for example, to appear as new traces rather than continuations of the originating trace.
I've tried extracting the header values and then setting appropriate values as Metadata on my grpc client calls but I haven't found sufficient information/API to tie things together properly.

I've just updated to `0.95.0` and am seeing the same behavior.  It does sound similar to the behavior described in this [reported issue](https://github.com/DataDog/dd-trace-java/issues/1633) but I'm using scala and Alpakka so I'm reporting it separately.
Is this the expected behavior, ie that a new trace is created for a `kafka.conusme` operation?

**Comment by mcculls** [comment]
Hi Gerrick,

The trace and span ids recorded in the `ConsumerRecord` are only extracted and used as the context at the point where we create the `kafka.consume` span - before that point the thread's MDC and global tracer will not reflect the record's context because it has yet to be extracted.

For the client API we try to create `kafka.consume` spans at the point where applications iterate over records rather than during internal processing because we want to capture the consume time from the application's perspective. For the streaming API we try to create `kafka.consume` spans just before the stream's processor node calls back into the application, for the same reason.

Depending where you've added your custom instrumentation it's possible that you'll see trace and span ids in the record that aren't yet reflected in the MDC or global tracer, for example before the `kafka.consume` span is created - but also after that span has finished, because it will reset the context back to the original value. If that is the case then you'd need to extract the trace and span id yourself from the record.

If your custom instrumentation is expected to be inside the `kafka.consume` span then we'd need more details about where you've added that to investigate further.

**Comment by mcculls** [comment]
thanks for the example - I see that Alpakka-kafka builds on akka-streams which we don't yet support (we also don't have support for web-socket propagation)

you should be able to use manual instrumentation to propagate the trace: https://docs.datadoghq.com/tracing/setup_overview/custom_instrumentation/java/#adding-spans - you can use that to get the current trace details from inside the consumer, you'd then need to decide how to propagate it over the websocket (whether using headers or part of the payload) and then build and activate a span on the other side using the trace details you just sent

**Comment by gbiv** [comment]
Got it!
TL;DR get the tracer and use an implementation of `TextMap` to extract the `Context` to use as the parent span. 

 What I didn't understand was how to create a `Context` from the existing information that was available in the `ConsumerRecord` `Headers`.  I eventually found [this issue](https://github.com/DataDog/dd-trace-java/issues/1815#issuecomment-683884708) which lead me to [the kinesis example ](https://github.com/DataDog/dd-trace-java-kinesis-example)which gave me the answer I was looking for.

I think it would be useful to mention that example( or something similar)  in the documentation around [manually creating spans](https://docs.datadoghq.com/tracing/setup_overview/custom_instrumentation/java/#manually-creating-a-new-span). Once I found that example, it was pretty straightforward.
